File: mac/PabloCompanion/Services/PHISanitizer.swift
import Foundation
import os

/// Strips Protected Health Information (PHI) from text before sending to the LLM.
///
/// The LLM only needs navigation structure — not clinical content.
/// This sanitizer removes patient names, phone numbers, emails, dates of birth,
/// SSNs, and ICD-10 diagnosis codes.
enum PHISanitizer {
    private static let logger = Logger(
        subsystem: AppConstants.appBundleID, category: "PHISanitizer"
    )

    /// Strips PHI from a DOM snapshot or page text.
    static func strip(from text: String, patientName: String) -> String {
        var stripped = text

        // Strip patient name and individual name parts
        stripped = stripped.replacingOccurrences(of: patientName, with: "[PATIENT]")
        for part in patientName.split(separator: " ") where part.count > 2 {
            stripped = stripped.replacingOccurrences(of: String(part), with: "[NAME]")
        }

        // Strip common PHI patterns
        for (pattern, replacement) in phiPatterns {
            if let regex = try? NSRegularExpression(pattern: pattern) {
                let range = NSRange(stripped.startIndex..., in: stripped)
                stripped = regex.stringByReplacingMatches(
                    in: stripped, range: range, withTemplate: replacement
                )
            }
        }

        return stripped
    }

    // Phone: (xxx) xxx-xxxx, xxx-xxx-xxxx, xxx.xxx.xxxx
    // Email: user@domain.com
    // DOB: MM/DD/YYYY, MM-DD-YYYY
    // SSN: xxx-xx-xxxx
    // ICD-10: letter + 2 digits + optional .digits
    private static let phiPatterns: [(String, String)] = [
        (#"\(?\d{3}\)?[-.\s]?\d{3}[-.\s]?\d{4}"#, "[PHONE]"),
        (#"[a-zA-Z0-9._%+-]+@[a-zA-Z0-9.-]+\.[a-zA-Z]{2,}"#, "[EMAIL]"),
        (#"\d{1,2}[/-]\d{1,2}[/-]\d{2,4}"#, "[DATE]"),
        (#"\d{3}-\d{2}-\d{4}"#, "[SSN]"),
        (#"\b[A-Z]\d{2}\.?\d{0,4}\b"#, "[DX]"),
    ]
}

/// Validates that an LLM-returned CSS selector is safe to inject into JavaScript.
///
/// Rejects selectors containing code injection patterns (javascript:, eval, fetch, etc.)
/// and selectors that are unreasonably long.
enum SelectorValidator {
    private static let logger = Logger(
        subsystem: AppConstants.appBundleID, category: "SelectorValidator"
    )

    /// Validates a selector. Throws if it contains forbidden patterns.
    static func validate(_ selector: String) throws {
        let lowered = selector.lowercased()
        for pattern in forbiddenPatterns where lowered.contains(pattern) {
            logger.error("Rejected unsafe selector: \(selector.prefix(100))")
            throw EHRNavigatorError.actionFailed(
                action: "validate",
                selector: "Selector rejected: contains forbidden pattern"
            )
        }
        if selector.count > 500 {
            throw EHRNavigatorError.actionFailed(
                action: "validate",
                selector: "Selector too long (\(selector.count) chars)"
            )
        }
    }

    private static let forbiddenPatterns = [
        "javascript:", "document.cookie", "document.location",
        "window.location", "fetch(", "xmlhttprequest",
        "eval(", "function(", "settimeout(", "setinterval(",
        "<script", "onerror=", "onload=", "onclick=",
    ]
}

File: windows/PabloCompanion/Services/PhiSanitizer.cs
using System.Text.RegularExpressions;

namespace PabloCompanion.Services;

/// <summary>
/// Strips Protected Health Information (PHI) from text before sending to the LLM.
/// The LLM only needs navigation structure — not clinical content.
/// Mirrors PHISanitizer.swift on macOS.
/// </summary>
public static partial class PhiSanitizer
{
    // Phone: (xxx) xxx-xxxx, xxx-xxx-xxxx, xxx.xxx.xxxx
    [GeneratedRegex(@"\(?\d{3}\)?[-.\s]?\d{3}[-.\s]?\d{4}")]
    private static partial Regex PhoneRegex();

    // Email: user@domain.com
    [GeneratedRegex(@"[a-zA-Z0-9._%+\-]+@[a-zA-Z0-9.\-]+\.[a-zA-Z]{2,}")]
    private static partial Regex EmailRegex();

    // DOB: MM/DD/YYYY, MM-DD-YYYY
    [GeneratedRegex(@"\d{1,2}[/\-]\d{1,2}[/\-]\d{2,4}")]
    private static partial Regex DateRegex();

    // DOB: YYYY-MM-DD (ISO format)
    [GeneratedRegex(@"\d{4}-\d{2}-\d{2}")]
    private static partial Regex IsoDateRegex();

    // SSN: xxx-xx-xxxx
    [GeneratedRegex(@"\b\d{3}-\d{2}-\d{4}\b")]
    private static partial Regex SsnRegex();

    // SSN: xxxxxxxxx (no dashes)
    [GeneratedRegex(@"\b\d{9}\b")]
    private static partial Regex SsnNoDashRegex();

    // MRN: MRN-123456, MRN: 123456, MRN#123456
    [GeneratedRegex(@"\bMRN[:\-\s#]*\d{4,10}\b", RegexOptions.IgnoreCase)]
    private static partial Regex MrnRegex();

    // ICD-10: letter + 2 digits + optional .digits
    [GeneratedRegex(@"\b[A-Z]\d{2}\.?\d{0,4}\b")]
    private static partial Regex Icd10Regex();

    /// <summary>
    /// Strips PHI from a DOM snapshot or page text.
    /// </summary>
    public static string Strip(string text, string patientName)
    {
        var stripped = text;

        // Strip patient full name
        stripped = stripped.Replace(patientName, "[PATIENT]", StringComparison.OrdinalIgnoreCase);

        // Strip individual name parts (> 2 chars)
        foreach (var part in patientName.Split(' ', StringSplitOptions.RemoveEmptyEntries))
        {
            if (part.Length > 2)
            {
                stripped = stripped.Replace(part, "[NAME]", StringComparison.OrdinalIgnoreCase);
            }
        }

        // Strip common PHI patterns
        stripped = SsnRegex().Replace(stripped, "[SSN]");
        stripped = SsnNoDashRegex().Replace(stripped, "[SSN]");
        stripped = MrnRegex().Replace(stripped, "[MRN]");
        stripped = PhoneRegex().Replace(stripped, "[PHONE]");
        stripped = EmailRegex().Replace(stripped, "[EMAIL]");
        stripped = DateRegex().Replace(stripped, "[DATE]");
        stripped = IsoDateRegex().Replace(stripped, "[DATE]");
        stripped = Icd10Regex().Replace(stripped, "[DX]");

        return stripped;
    }
}

/// <summary>
/// Validates that a CSS selector is safe to inject into JavaScript.
/// Rejects selectors containing code injection patterns.
/// Mirrors SelectorValidator in PHISanitizer.swift on macOS.
/// </summary>
public static class SelectorValidator
{
    private static readonly string[] ForbiddenPatterns =
    [
        "javascript:", "document.cookie", "document.location",
        "window.location", "fetch(", "xmlhttprequest",
        "eval(", "function(", "settimeout(", "setinterval(",
        "<script", "onerror=", "onload=", "onclick=",
    ];

    /// <summary>
    /// Validates a selector. Throws if it contains forbidden patterns or is too long.
    /// </summary>
    public static void Validate(string selector)
    {
        var lowered = selector.ToLowerInvariant();
        foreach (var pattern in ForbiddenPatterns)
        {
            if (lowered.Contains(pattern, StringComparison.Ordinal))
            {
                throw new EhrNavigatorException(
                    $"Selector rejected: contains forbidden pattern '{pattern}'");
            }
        }

        if (selector.Length > 500)
        {
            throw new EhrNavigatorException(
                $"Selector too long ({selector.Length} chars)");
        }
    }
}
